File: packages/contracts/src/content.ts
import { z } from "zod";

export const ChannelSchema = z.enum([
  "INSTAGRAM",
  "FACEBOOK",
  "LINKEDIN",
  "X",
  "REDDIT",
  "BLOG",
  "EMAIL",
  "GENERIC",
]);

export const ContentFormatSchema = z.enum([
  "POST",
  "CAROUSEL",
  "STORY",
  "SHORT_VIDEO",
  "LONG_FORM",
  "COMMENT",
  "THREAD",
]);

export const DraftStatusSchema = z.enum([
  "GENERATED",
  "PENDING_REVIEW",
  "APPROVED",
  "REJECTED",
  "SUPERSEDED",
]);

export const ContentEvidenceItemSchema = z.object({
  type: z.enum([
    "BRAND_FACT",
    "COMMERCE_METRIC",
    "RESEARCH_FINDING",
    "OPPORTUNITY",
    "OWNER_HINT",
  ]),
  ref: z.string(),
  summary: z.string(),
});

export const CarouselSlideSchema = z.object({
  slideNumber: z.number().int().positive(),
  text: z.string(),
  visualDirection: z.string().optional(),
});

// Flexible generated content — NestJS validates channel-specific fields at service level.
export const GeneratedContentSchema = z.object({
  channel: ChannelSchema,
  format: ContentFormatSchema,
  // Instagram / general
  caption: z.string().optional(),
  callToAction: z.string().optional(),
  hashtags: z.array(z.string()).optional(),
  // Instagram carousel
  hookSlide: z.string().optional(),
  slides: z.array(CarouselSlideSchema).optional(),
  closingCta: z.string().optional(),
  // X
  text: z.string().optional(),
  thread: z.array(z.string()).optional(),
  // LinkedIn
  // (text field reused)
  // Reddit
  title: z.string().optional(),
  body: z.string().optional(),
  subredditSuggestion: z.string().optional(),
  // Blog
  outline: z.array(z.string()).optional(),
  metaDescription: z.string().optional(),
  // Weave creative direction (M8)
  creativeDirection: z
    .object({
      aspectRatio: z.string().optional(),
      visualObjective: z.string().optional(),
      mood: z.string().optional(),
      requiredElements: z.array(z.string()).optional(),
      forbiddenElements: z.array(z.string()).optional(),
      productRefs: z.array(z.string()).optional(),
      textHierarchy: z.array(z.string()).optional(),
    })
    .optional(),
});

export const CriticEvaluationSchema = z.object({
  brandFit: z.number().min(0).max(1),
  channelFit: z.number().min(0).max(1),
  evidenceAlignment: z.number().min(0).max(1),
  clarity: z.number().min(0).max(1),
  originality: z.number().min(0).max(1),
  promotionalIntensity: z.number().min(0).max(1),
  claimRisk: z.number().min(0).max(1),
  ctaQuality: z.number().min(0).max(1),
  overall: z.number().min(0).max(1),
  issues: z.array(z.string()),
  passesReview: z.boolean(),
});

const ContentBriefSummarySchema = z.object({
  objective: z.string(),
  topic: z.string(),
  angle: z.string(),
  targetAudience: z.string(),
  channel: ChannelSchema,
  format: ContentFormatSchema,
  keyMessage: z.string(),
  callToAction: z.string().optional(),
  tone: z.string(),
  constraints: z.array(z.string()),
});

const ContentBrandContextSchema = z.object({
  name: z.string(),
  voice: z.string().optional(),
  audience: z.string().optional(),
  guidelines: z.array(
    z.object({
      category: z.string(),
      rule: z.string(),
      example: z.string().optional(),
    }),
  ),
  activeProducts: z.array(
    z.object({
      name: z.string(),
      description: z.string().optional(),
      category: z.string().optional(),
      tags: z.array(z.string()),
    }),
  ),
});

const ContentEvidenceSchema = z.object({
  // Trusted sources
  brandFacts: z.array(z.string()),
  commerceSummary: z.string().optional(),
  // Untrusted external research — labeled clearly in prompt
  researchFindings: z.array(z.string()),
  opportunitySummary: z.string().optional(),
  ownerHint: z.string().optional(),
});

export const ContentGenerationRequestSchema = z.object({
  brief: ContentBriefSummarySchema,
  brandContext: ContentBrandContextSchema,
  evidence: ContentEvidenceSchema,
  revisionFeedback: z.string().optional(),
});

export const ContentCriticRequestSchema = z.object({
  content: GeneratedContentSchema,
  brief: ContentBriefSummarySchema,
  brandContext: ContentBrandContextSchema,
});

export type Channel = z.infer<typeof ChannelSchema>;
export type ContentFormat = z.infer<typeof ContentFormatSchema>;
export type DraftStatus = z.infer<typeof DraftStatusSchema>;
export type ContentEvidenceItem = z.infer<typeof ContentEvidenceItemSchema>;
export type CarouselSlide = z.infer<typeof CarouselSlideSchema>;
export type GeneratedContent = z.infer<typeof GeneratedContentSchema>;
export type CriticEvaluation = z.infer<typeof CriticEvaluationSchema>;
export type ContentGenerationRequest = z.infer<
  typeof ContentGenerationRequestSchema
>;
export type ContentCriticRequest = z.infer<typeof ContentCriticRequestSchema>;

File: apps/brain/app/schemas/content.py
from pydantic import BaseModel, field_validator
from typing import Optional, List, Literal
from enum import Enum


class Channel(str, Enum):
    INSTAGRAM = "INSTAGRAM"
    FACEBOOK = "FACEBOOK"
    LINKEDIN = "LINKEDIN"
    X = "X"
    REDDIT = "REDDIT"
    BLOG = "BLOG"
    EMAIL = "EMAIL"
    GENERIC = "GENERIC"


class ContentFormat(str, Enum):
    POST = "POST"
    CAROUSEL = "CAROUSEL"
    STORY = "STORY"
    SHORT_VIDEO = "SHORT_VIDEO"
    LONG_FORM = "LONG_FORM"
    COMMENT = "COMMENT"
    THREAD = "THREAD"


class CarouselSlide(BaseModel):
    slideNumber: int
    text: str
    visualDirection: Optional[str] = None


class CreativeDirection(BaseModel):
    aspectRatio: Optional[str] = None
    visualObjective: Optional[str] = None
    mood: Optional[str] = None
    requiredElements: Optional[List[str]] = None
    forbiddenElements: Optional[List[str]] = None
    productRefs: Optional[List[str]] = None
    textHierarchy: Optional[List[str]] = None


class GeneratedContent(BaseModel):
    channel: Channel
    format: ContentFormat
    # Shared / general
    caption: Optional[str] = None
    callToAction: Optional[str] = None
    hashtags: Optional[List[str]] = None
    # Instagram carousel
    hookSlide: Optional[str] = None
    slides: Optional[List[CarouselSlide]] = None
    closingCta: Optional[str] = None
    # X / Reddit / LinkedIn text
    text: Optional[str] = None
    thread: Optional[List[str]] = None
    # Reddit
    title: Optional[str] = None
    body: Optional[str] = None
    subredditSuggestion: Optional[str] = None
    # Blog
    outline: Optional[List[str]] = None
    metaDescription: Optional[str] = None
    # Weave creative direction (M8)
    creativeDirection: Optional[CreativeDirection] = None


class CriticEvaluation(BaseModel):
    brandFit: float
    channelFit: float
    evidenceAlignment: float
    clarity: float
    originality: float
    promotionalIntensity: float
    claimRisk: float
    ctaQuality: float
    overall: float
    issues: List[str]
    passesReview: bool

    @field_validator(
        "brandFit",
        "channelFit",
        "evidenceAlignment",
        "clarity",
        "originality",
        "promotionalIntensity",
        "claimRisk",
        "ctaQuality",
        "overall",
    )
    @classmethod
    def score_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("score must be between 0.0 and 1.0")
        return v


# --- Request / response schemas for brain endpoints ---

class ContentBriefRequest(BaseModel):
    objective: str
    topic: str
    angle: str
    targetAudience: str
    channel: Channel
    format: ContentFormat
    keyMessage: str
    callToAction: Optional[str] = None
    tone: str
    constraints: List[str] = []


class ContentBrandContext(BaseModel):
    name: str
    voice: Optional[str] = None
    audience: Optional[str] = None
    guidelines: List[dict] = []
    activeProducts: List[dict] = []


class ContentEvidence(BaseModel):
    brandFacts: List[str] = []
    commerceSummary: Optional[str] = None
    researchFindings: List[str] = []
    opportunitySummary: Optional[str] = None
    ownerHint: Optional[str] = None


class ContentGenerationRequest(BaseModel):
    brief: ContentBriefRequest
    brandContext: ContentBrandContext
    evidence: ContentEvidence
    revisionFeedback: Optional[str] = None


class ContentCriticRequest(BaseModel):
    content: GeneratedContent
    brief: ContentBriefRequest
    brandContext: ContentBrandContext
